extensions/amp-next-page/1.0/visibility-observer.js

```diff
@@ -66,86 +95,68 @@ export class VisibilityObserverEntry {
     this.observer_
       .getPositionObserver()
       .observe(top, PositionObserverFidelity.LOW, position =>
-        this.topSentinelPositionChanged(position, callback)
+        this.topSentinelPositionChanged(position)
       );
     this.observer_
       .getPositionObserver()
       .observe(bottom, PositionObserverFidelity.LOW, position =>
-        this.bottomSentinelPositionChanged(position, callback)
+        this.bottomSentinelPositionChanged(position)
       );
   }
 
   /**
    * Called when a position change is detected on the injected
    * top sentinel element
    * @param {?PositionInViewportEntryDef} position
-   * @param {function(!ViewportRelativePos)} callback
    */
-  topSentinelPositionChanged(position, callback) {
-    const prevTopSentinelPosition = this.topSentinelPosition_;
-    if (!position || position.relativePos === prevTopSentinelPosition) {
-      return;
-    }
+  topSentinelPositionChanged(position) {
     this.topSentinelPosition_ = position.relativePos;
-    this.updateRelativePos_(callback);
+    this.observer_.updateRelativePos(this);
   }
 
   /**
    * Called when a position change is detected on the injected
    * bottom sentinel element
    * @param {?PositionInViewportEntryDef} position
-   * @param {function(!ViewportRelativePos)} callback
    */
-  bottomSentinelPositionChanged(position, callback) {
-    const prevBottomSentinelPosition = this.bottomSentinelPosition_;
-    if (!position || position.relativePos === prevBottomSentinelPosition) {
-      return;
-    }
+  bottomSentinelPositionChanged(position) {
     this.bottomSentinelPosition_ = position.relativePos;
-    this.updateRelativePos_(callback);
+    this.observer_.updateRelativePos(this);
   }
 
   /**
-   * Calculates the position of the element relative to the viewport
-   * based on the positions of the injected bottom and top sentinel elements
+   * @return {boolean}
+   */
+  usesSentinel() {
+    return true;
+  }
+}
+
+class VisibilityObserverHostEntry extends VisibilityObserverEntry {
```

Comment:
Should we extend or compose here instead? Is there precedence for extending this?

Maybe we can just use public methods of an inner `VisibilityObserverEntry`.

Reply:
extending feels cleaner to me but open to discussion

Reply:
In my experience extension hierarchies tend to complicate things over time.

Why not invert the model?

- use a single `VisibilityObserverEntry` type
- inject the dependencies of implementation areas that change. These can be either:
  - two different objects with the same interface (like autocomplete or viewport), or
  - granular callbacks injected on instantiation

Shared activities would be handled at the higher `VisibilityObserverEntry` level and make them more explicit.

So for example, rather than conditionally checking `usesSentinel()`, you'd instead inject the sentinel behavior. This simplifies because you don't have to know what a `sentinel`, or other local concepts are when you don't need them, and the implementation flow goes in one direction.

LMK if you'd like to discuss concrete details offline.

Reply:
Since this PR is critical for a P1 bug #27635 , let's keep it as is and do the refactor in a separate PR as to not introduce any new issues. Opened https://github.com/ampproject/amphtml/issues/27636 to track this.